flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/DataType.java

```diff
@@ -0,0 +1,427 @@
+		public boolean equals(Object o) {
+			if (this == o) {
+				return true;
+			}
+			if (o == null || getClass() != o.getClass()) {
+				return false;
+			}
+			if (!super.equals(o)) {
+				return false;
+			}
+			ElementDataType that = (ElementDataType) o;
+			return elementDataType.equals(that.elementDataType);
+		}
+
+		@Override
+		public int hashCode() {
+			return Objects.hash(super.hashCode(), elementDataType);
+		}
+	}
+
+	/**
+	 * A data type that contains a key and value data type (e.g. {@code MAP}).
+	 *
+	 * @see DataTypes for a list of supported data types
+	 */
+	public static final class KeyValueDataType extends DataType {
+
+		private final DataType keyDataType;
+
+		private final DataType valueDataType;
+
+		public KeyValueDataType(
+				LogicalType logicalType,
+				@Nullable Class<?> conversionClass,
+				DataType keyDataType,
+				DataType valueDataType) {
+			super(logicalType, conversionClass);
+			this.keyDataType = Preconditions.checkNotNull(keyDataType, "Key data type must not be null.");
+			this.valueDataType = Preconditions.checkNotNull(valueDataType, "Value data type must not be null.");
+		}
+
+		public KeyValueDataType(
+				LogicalType logicalType,
+				DataType keyDataType,
+				DataType valueDataType) {
+			this(logicalType, null, keyDataType, valueDataType);
+		}
+
+		public DataType getKeyDataType() {
+			return keyDataType;
+		}
+
+		public DataType getValueDataType() {
+			return valueDataType;
+		}
+
+		@Override
+		public DataType notNull() {
+			return new KeyValueDataType(
+				logicalType.copy(false),
+				conversionClass,
+				keyDataType,
+				valueDataType);
+		}
+
+		@Override
+		public DataType andNull() {
+			return new KeyValueDataType(
+				logicalType.copy(true),
+				conversionClass,
+				keyDataType,
+				valueDataType);
+		}
+
+		@Override
+		public DataType bridgedTo(Class<?> newConversionClass) {
+			return new KeyValueDataType(
+				logicalType,
+				Preconditions.checkNotNull(newConversionClass, "New conversion class must not be null."),
+				keyDataType,
+				valueDataType);
+		}
+
+		@Override
+		public boolean equals(Object o) {
+			if (this == o) {
+				return true;
+			}
+			if (o == null || getClass() != o.getClass()) {
+				return false;
+			}
+			if (!super.equals(o)) {
+				return false;
+			}
+			KeyValueDataType that = (KeyValueDataType) o;
+			return keyDataType.equals(that.keyDataType) && valueDataType.equals(that.valueDataType);
+		}
+
+		@Override
+		public int hashCode() {
+			return Objects.hash(super.hashCode(), keyDataType, valueDataType);
+		}
+	}
+
+	/**
+	 * A data type that contains field data types (e.g. {@code ROW} or structured types).
+	 *
+	 * @see DataTypes for a list of supported data types
+	 */
+	public static final class FieldsDataType extends DataType {
+
+		private final Map<String, DataType> fieldDataTypes;
+
+		public FieldsDataType(
+				LogicalType logicalType,
+				@Nullable Class<?> conversionClass,
+				Map<String, DataType> fieldDataTypes) {
```

Comment:
It's a bit strange to use `Map<String, DataType>` to represent fields, because RowType is sequential, is it better to use array to express fields?

Reply:
I agree. At least we should use LinkedHashMap to preserve the fields order. And it would be nice to provide `getFieldNames()` and `getFieldTypes()`  for convenience.

Reply:
The field order is defined by the logical type. The map adds just metadata.

Reply:
If we would allow a `LinkedHashMap` we would also need to verify the order with the logical type. The logical type should be the single source of truth.

Reply:
OK, I'm just not sure what the `getFieldDataTypes` will be used for.

Reply:
This method will only used by us to figure out which format the user requested. It will be used in  the next PR around FLINK-12254.